lib/resty/upstream/healthcheck.lua

```diff
@@ -613,14 +704,21 @@ function _M.spawn_checker(opts)
         return nil, "no upstream specified"
     end
 
-    local ppeers, err = get_primary_peers(u)
-    if not ppeers then
-        return nil, "failed to get primary peers: " .. err
-    end
+    local dedup = opts.dedup
+
+    local ppeers, bpeers
+    if dedup then
+        ppeers, bpeers = get_all_unchecked_peers(u)
```

Comment:
This is the only place you're calling `get_all_unchecked_peers`, and that function is the only caller of `get_unchecked_peers`. Here you're setting `checked_peers_upstream[peer.name] = u` for all unchecked peers. It seems more succinct to:
```lua
local ppeers, bpeers = get_all_peers(u)
if dedup then
  ppeers = select_and_check_unchecked_peers(ppeers, u)
  bpeers = select_and_check_unchecked_peers(bpeers, u)
end
```
here, with:
```lua
local function select_and_check_unchecked_peers(peers, u)
  local filtered_peers = {}
  for _, peer in ipairs(peers) do
    if not checked_peers_upstream[peer.name] then
      checked_peers_upstream[peer.name] = u
      table.insert(filtered_peers, peer)
    end
  end
  return filtered_peers
end
```
If there's overlap between backup and primary peer names (e.g. some primary peers appear as backup peers for other upstreams), then the result is slightly different from your code. If the distinction is important, it's easy to modify `select_and_check_unchecked_peers` to take `(ppeers, bpeers, u)` and loop over both arrays, filtering first, then setting.